File: 02-week-2--recursion/04-day-4--recursive-find-shortest-string/ruby/find_shortest_string_recursive.rb
def find_shortest_string_recursive(arr)
  # type your code in here
  return arr[0] if arr.length == 1

  second_word = find_shortest_string_recursive(arr[1..-1])

  arr[0].length <= second_word.length ? arr[0] : second_word
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 'a'"
  puts "=>", find_shortest_string_recursive(['aaa', 'a', 'bb', 'ccc'])

  puts

  puts "Expecting: 'hi'"
  puts "=>", find_shortest_string_recursive(['cat', 'hi', 'dog', 'an'])

  puts

  puts "Expecting: 'lily'"
  puts "=>", find_shortest_string_recursive(['flower', 'juniper', 'lily', 'dandelion'])

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution

#Iterative solution

# def find_shortest_string_recursive

#arr.sort_by{|s| s.length }[0]

#end

File: 02-week-2--recursion/04-day-4--recursive-find-shortest-string/javascript/find_shortest_string_recursive.js
function findShortestStringRecursive(arr) {
  // type your code here
  if (arr.length === 1){
    return arr[0]
  }

  const secondWord = findShortestStringRecursive(arr.slice(1))

  return arr[0].length <= secondWord.length ? arr[0] : secondWord
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 'a'");
  console.log("=>", findShortestStringRecursive(['aaa', 'a', 'bb', 'ccc']));

  console.log("");

  console.log("Expecting: 'hi'");
  console.log("=>", findShortestStringRecursive(['cat', 'hi', 'dog', 'an']));

  console.log("");

  console.log("Expecting: 'lily'");
  console.log("=>", findShortestStringRecursive(['flower', 'juniper', 'lily', 'dandelion']));
}

module.exports = findShortestStringRecursive;

// Please add your pseudocode to this file
// And a written explanation of your solution

//Solution without recursion
//function findShortestString(arr){

// return arr.sort((a,b) => a.length - b.length)[0]

// }